core/unit_test/TestHostSharedPtrAccessOnDevice.hpp

```diff
@@ -157,3 +157,123 @@ TEST(TEST_CATEGORY, host_shared_ptr_special_members_on_device) {
   check_special_members_on_device(device_ptr);
 }
 #endif
+
+namespace {
+
+struct Bar {
+  double val;
+};
+
+struct Foo {
+  Foo(bool allocate = false) : ptr(allocate ? new Bar : nullptr) {}
+  Kokkos::Impl::HostSharedPtr<Bar> ptr;
+  int use_count() { return (int)ptr.use_count(); }
+};
+
+template <class DevMemSpace, class HostMemSpace>
+void host_shared_ptr_test_reference_counting() {
+  using ExecSpace = typename DevMemSpace::execution_space;
+  bool is_gpu     = ExecSpace().concurrency() > 10000;
```

Comment:
Forward declare, and `std::is_same<ExecSpace, Cuda> || ...`. This just begs for some architecture where 100k threads can exist to screw us.